bindsnet/learning/__init__.py

```diff
@@ -194,7 +196,7 @@ def __init__(self, connection: AbstractConnection, nu: Optional[Union[float, Seq
         )
 
         assert self.source.traces, 'Pre-synaptic nodes must record spike traces.'
-        assert connection.wmin is not None and connection.wmax is not None, 'Connection must define wmin and wmax.'
+        assert connection.wmin != -np.inf and connection.wmax != np.inf, 'Connection must define finite wmin and wmax.'
```

Comment:
We allow infinite weights. This assert should be removed.

Reply:
Yes we do in general, but for this specific learning rule (which is based on the difference between weight bounds and actual weights), they should be finite right? Otherwise weights would become infinite here:
https://github.com/Hananel-Hazan/bindsnet/blob/f640fb602c6b550b4c9858ee533848fc3c02f135/bindsnet/learning/__init__.py#L227

Reply:
No, I think the weights should be able to get arbitrarily large. They won't actually reach `np.inf` / `-np.inf` since that's impossible.

Reply:
Ok, so say the assert would be removed, and `self.wmin` would be set to `connection.wmin`, which would be `-np.inf`. Then, when learning, `self.update` would be called (which would be `self._connection_update` for a `Connection`) and a proposed `update` to the weights would be computed based on the difference `self.connection.w - self.wmin`. The `update` would become `inf` because the difference is infinite. And, therefore, `self.connection.w` would become `inf`.

According to this, this update rule just won't work without the weight bounds, and it maybe would be better to let people know, rather than let them find out their weights have gone to infinity.

If anything is wrong with the above, please explain because I really don't see it :sweat_smile:

Reply:
Ah, okay, that makes sense. I think this is all set to merge, then.